### PR TITLE
APEXCORE-304 Added support for libjars in local mode.

### PR DESCRIPTION
**Pull request by chinmaykolhatkar**
@PramodSSImmaneni Please review

### CODE REVIEW

**Comment by chinmaykolhatkar** [comment]
Not sure why the travis build is failing.. Travis command is passing successfully when run locally.

File: api/src/main/java/com/datatorrent/api/StreamingApplication.java [hunk 1]
```diff
@@ -47,6 +47,12 @@
   String ENVIRONMENT = DT_PREFIX + "environment";
 
   /**
+   * Use this config variable to to add any external library to classpath for the DAG.
+   * Value of this variable is same as StramAppLauncher.LIBJARS_CONF_KEY_NAME.
+   */
+  String LIBJARS_CONF_KEY_NAME = "tmplibjars";
```
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/StreamingApplication.java]: I don't think this belongs into the API.

**Comment by tweise** [comment]
The ticket seems to suggest a broader change than just local mode. What am I missing?

File: api/src/main/java/com/datatorrent/api/StreamingApplication.java (hunk 1 above)
Comment (chinmaykolhatkar) [inline api/src/main/java/com/datatorrent/api/StreamingApplication.java]: The reason why it's added here is constant for tmplibjars is present in StramLauncher which is part of apex-engine and if this variable is to be used in malhar, apex-engine won't be added as dependency to malhar.

**Comment by chinmaykolhatkar** [comment]
The idea behind this ticket is to allow user to provide external jars to classpath.

We already do have a way to provide external jars using tmplibjars conf property.
But this works only for cluster mode and not for local mode.
This change change makes local mode also consume tmplibjars conf variable.

This way the purpose of served without creating extra APIs etc...

I can change the description of the ticket is that's right to do.

**Comment by tweise** [comment]
Yes, the ticket should reflect the actual work being done.

File: api/src/main/java/com/datatorrent/api/StreamingApplication.java (hunk 1 above)
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/StreamingApplication.java]: Just looking at the string literal suggests something isn't right here. We configure applications through attributes, not random properties in the config file. Can you please check if there is a better way to express this dependency.

Comment (chinmaykolhatkar) [inline api/src/main/java/com/datatorrent/api/StreamingApplication.java]: Hmmm..  got your point and agree with you..  I find attribute as a better place than properties for such configuration. 

Here is my suggestion about exposing this as an attribute:
1. Have a attribute in DAGContext:
Attribute<Map<String, Boolean>> APP_EXTERNAL_JARS = new Attribute<Map<, >(new Map2String<String, Boolean>(",", "=", new String2String<>(), new String2String<>()));
1. We pick up this attribute in following places:
   a. For cluster mode: StramAppLauncher.launchApp
   b. For local mode (apex cli) : StramAppLauncher.runLocal
   c. For local mode (ApplicationTest): LocalModeImpl.getController
2. For individual approaches, jar becomes available from that point onwards.

Can you share though on above approach?

Comment (chinmaykolhatkar) [inline api/src/main/java/com/datatorrent/api/StreamingApplication.java]: Hmmm..  got your point and agree with you..  I find attribute as a better place than properties for such configuration. 

Here is my suggestion about exposing this as an attribute:
1. Have a attribute in DAGContext:
Attribute<Map<String, Boolean>> APP_EXTERNAL_JARS = new Attribute<Map<, >(new Map2String<String, Boolean>(",", "=", new String2String<>(), new String2String<>()));
1. We pick up this attribute in following places:
   a. For cluster mode: StramAppLauncher.launchApp
   b. For local mode (apex cli) : StramAppLauncher.runLocal
   c. For local mode (ApplicationTest): LocalModeImpl.getController
2. For individual approaches, jar becomes available from that point onwards.

Can you share though on above approach?

File: engine/src/test/java/com/datatorrent/stram/StramLocalClusterTest.java [hunk 2]
```diff
@@ -274,4 +286,103 @@ public WindowGenerator setupWindowGenerator()
...
+        dynamicLoader.setClass("POJO");
+
+        dag.addStream("fromNode1", genNode.outport, dynamicLoader.in);
+
+        dag.setInputPortAttribute(dynamicLoader.in, Context.PortContext.TUPLE_CLASS, pojo);
+        conf.set(LIBJARS_CONF_KEY_NAME, generatedJar);
+      }
+    };
+
+    LocalMode.runApp(app, 10000);
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/StramLocalClusterTest.java]: Please structure the unit tests so that you wait for expected result instead of waiting for arbitrary time. There are examples in Malhar for this.

File: engine/src/test/java/com/datatorrent/stram/StramLocalClusterTest.java [hunk 3]
```diff
@@ -274,4 +286,103 @@ public WindowGenerator setupWindowGenerator()
...
+
+    FileUtils.forceDelete(new File("src/test/resources/dynamicJar/POJO.class"));
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/StramLocalClusterTest.java]: Why is the unit test (creating? and) deleting files in the source tree?

File: engine/src/test/java/com/datatorrent/stram/StramLocalClusterTest.java [hunk 4]
```diff
@@ -274,4 +286,103 @@ public WindowGenerator setupWindowGenerator()
...
+    };
+
+    @Override
+    public void setup(Context.OperatorContext context)
+    {
+      try {
+        Thread.currentThread().getContextClassLoader().loadClass(classToLoad);
+        Assert.assertTrue(true);
+      } catch (ClassNotFoundException e) {
+        Assert.fail(e.getMessage());
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/StramLocalClusterTest.java]: propagate exception

File: api/src/main/java/com/datatorrent/api/StreamingApplication.java (hunk 1 above)
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/StreamingApplication.java]: There already is an attribute that was taken out of the public API some time ago... LogicalPlan.LIBRARY_JARS

**Comment by chinmaykolhatkar** [comment]
All comments are taken care of.
@tweise @PramodSSImmaneni Please review.

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java [hunk 5]
```diff
@@ -2386,7 +2381,7 @@ public static void write(DAG dag, OutputStream os) throws IOException
 
   public static LogicalPlan read(InputStream is) throws IOException, ClassNotFoundException
   {
-    return (LogicalPlan)new ObjectInputStream(is).readObject();
+    return (LogicalPlan)new ClassLoaderObjectInputStream(Thread.currentThread().getContextClassLoader(), is).readObject();
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java:2384]: Why is this required (or how was it working prior to this change)?

File: engine/src/test/java/com/datatorrent/stram/StramLocalClusterTest.java [hunk 6]
```diff
@@ -274,4 +287,113 @@ public WindowGenerator setupWindowGenerator()
...
+
+      DynamicLoaderApp.passed = true;
+      DynamicLoaderApp.latch.countDown();
+    }
+  }
+
+  private String generatejar(String pojoClassName) throws IOException, InterruptedException
+  {
+    String sourceDir = "src/test/resources/dynamicJar/";
+    String destDir = "target/";
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/StramLocalClusterTest.java]: use StramTestSupport.TestMeta

Comment (chinmaykolhatkar) [inline engine/src/test/java/com/datatorrent/stram/StramLocalClusterTest.java]: Done.

File: engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java (hunk 5 above)
Comment (chinmaykolhatkar) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java:2384]: This is required for Local Mode case.

Lets say a jar path is set to  LIBRARY_JARS and this jar contains a POJO that needs to be set to TUPLE_CLASS attr. Now,  one can do that in populateDAG method, that would cause a problem in  LocalMode as follows:
1. populateDAG sets TUPLE_CLASS attr with new POJO class - Works.
2. Local mode serializes the LogicalPlan - Works.
3. Local mode deserialized the LogicalPlan - Fails.

3rd step fails becuase The ObjectInputStream deserializer uses System ClassLoader which does that have that POJO. ClassLoaderObjectInputStream is an overridden version of ObjectInputStream which allows to set the custom classloader.

By this time of callstack, Thread.currentThread().getClassLoader() will have a classloader which is super of all.

As previously, there was no functionality of LIBRARY_JARS, I guess this was not tested and so did not surface as an issue.

For cluster mode, there is no difference. i.e. There is only 1 classloader i.e. System.Classloader and that has new jars as well.
i.e. For cluster mode, Thread.currentThread().getClassLoader() = System classloader
i.e. ClassLoaderObjectInputStream will behave as if ObjectInputStream.
